src/vea/vea_free.c

```diff
@@ -159,21 +160,56 @@ merge_free_ext(struct vea_space_info *vsi, struct vea_free_extent *ext_in,
 
 	D_ASSERT(daos_handle_is_valid(btr_hdl));
 	d_iov_set(&key, &ext_in->vfe_blk_off, sizeof(ext_in->vfe_blk_off));
-repeat:
 	d_iov_set(&key_out, NULL, 0);
 	d_iov_set(&val, NULL, 0);
 
-	rc = dbtree_fetch(btr_hdl, opc, DAOS_INTENT_DEFAULT, &key, &key_out,
-			  &val);
-	if (rc == -DER_NONEXIST && opc == BTR_PROBE_LE) {
-		opc = BTR_PROBE_GE;
-		goto repeat;
+	/*
+	 * Search the one to be merged, the btree trace will be set to proper position for
+	 * later potential insertion (when merge failed), so don't change btree trace!
+	 */
+	rc = dbtree_fetch(btr_hdl, BTR_PROBE_EQ, DAOS_INTENT_DEFAULT, &key, &key_out, &val);
+	if (rc == 0) {
+		D_ERROR("unexpected extent ["DF_U64", %u]\n",
+			ext_in->vfe_blk_off, ext_in->vfe_blk_cnt);
+		return -DER_INVAL;
+	} else if (rc != -DER_NONEXIST) {
+		D_ERROR("search extent with offset "DF_U64" failed. "DF_RC"\n",
+			ext_in->vfe_blk_off, DP_RC(rc));
+		return rc;
 	}
+repeat:
+	d_iov_set(&key_out, NULL, 0);
+	d_iov_set(&val, NULL, 0);
 
-	if (rc == -DER_NONEXIST)
-		goto done;	/* Merge done */
-	else if (rc)
-		return rc;	/* Error */
+	if (fetch_prev) {
+		rc = dbtree_fetch_prev(btr_hdl, &key_out, &val, false);
+		if (rc == -DER_NONEXIST) {
+			fetch_prev = false;
+			goto repeat;
```

Comment:
If we cannot find the prev one, then can we set some flag (before goto repeat) to skip dbtree_fetch_cur() when try to find next?

Reply:
I see what happened under such case